### PR TITLE
[core][gcs] Add storage namespace to redis storage in GCS.

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?
To enable one storage be able to be shared by multiple ray clusters, a special prefix is added to isolate the data between clusters: "<EXTERNAL_STORAGE_NAMESPACE>@"

The namespace is given by an os environment: `RAY_external_storage_namespace` when start the head: `RAY_external_storage_namespace=1234 ray start --head`

This flag is very important in HA GCS environment. For example, in ray serve operator, when the operator tries to bring up a new one, it's hard to just start a new db, but it's relatively easy to generate a new cluster id.
Another example is that, the user might only be able to maintain one HA Redis DB, and the namespace enable the user to start multiple ray clusters which share the same db.

This config should be moved to storage config in the future once we build that.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by mwtian — COMMENTED** [review]
One question, what about creating a random unique cluster ID internally in Ray, when a head node starts? The advantages are less complexity when starting Ray, and no room for errors where kuberay / Anyscale product forget to set the cluster ID.

**Comment by iycheng** [comment]
@scv119 @mwtian do you have other comments?

**Review by mwtian — COMMENTED** [review]
For cluster ID,
- I agree for HA, it seems easier to allow the cluster ID to be set by kuberay etc.
- We probably should limit the ID to a well defined format, e.g. an integer, or fixed length hex characters.
- Should Ray use a random cluster ID by default and allow it to be overridden? Then sharing external Redis between clusters would work by default. But it is unsupported right now anyway.

Other than these questions and comments, the change looks good.

**Comment by iycheng** [comment]
@mwtian Thanks for the review.
I don't have strong opinion about this, but I think the direction we go should depend on the actual use case.

For 2), I'm not sure whether it's useful or just adds extra complexity. Any practical reason you think we should do this?

For 3) I am also not sure about this. Right now, we only have KubeRay trying to use this.

File: src/ray/common/ray_config_def.h [hunk 1]
```diff
@@ -612,6 +612,10 @@ RAY_CONFIG(bool, scheduler_avoid_gpu_nodes, true)
 /// Whether to skip running local GC in runtime env.
 RAY_CONFIG(bool, runtime_env_skip_local_gc, false)
 
+/// The ray cluster id. Different cluster should have unique ID.
+/// This fields is used to isolate data stored in DB.
+RAY_CONFIG(std::string, cluster_id, "default")
```
Comment (rkooo567) [inline src/ray/common/ray_config_def.h:617]: Isn't the default supposed to be an unique id if the different cluster should have unique id?

File: src/ray/gcs/store_client/redis_store_client.cc [hunk 2]
```diff
@@ -17,23 +17,143 @@
...
+
+std::string GenRedisKey(const std::string &cluster_id,
+                        const std::string &table_name,
+                        const std::string &key) {
+  return absl::StrCat(cluster_id, kClusterSeparator, table_name, kTableSeparator, key);
+}
+
+std::string GenKeyRedisMatchPattern(const std::string &cluster_id,
+                                    const std::string &table_name) {
+  return absl::StrCat(EscapeMatchPattern(cluster_id),
```
[the remark this answers is not in the document]
Comment (iycheng) [reply src/ray/gcs/store_client/redis_store_client.cc:49]: resolved offline.

**Comment by iycheng** [comment]
@mwtian @rkooo567 @scv119 it's ready for another review. Really appreciate the discussion.

File: src/ray/gcs/gcs_server/gcs_kv_manager.cc [hunk 3]
```diff
@@ -22,37 +22,45 @@
...
+  view = view.substr(cluster_id_.size() + kClusterSeparator.size());
+  if (absl::StartsWith(view, kNamespacePrefix)) {
     std::vector<std::string> parts =
         absl::StrSplit(key, absl::MaxSplits(kNamespaceSep, 1));
     RAY_CHECK(parts.size() == 2) << "Invalid key: " << key;
 
     return parts[1];
   }
-  return key;
+  return std::string(view.begin(), view.end());
```
[the remark this answers is not in the document]
Comment (mwtian) [reply src/ray/gcs/gcs_server/gcs_kv_manager.cc:64]: ^^?

**Review by mwtian — APPROVED** [review]
LGTM on the API, based on the off-thread discussion. I did not check the moved code in `src/ray/gcs/store_client/redis_store_client.cc`. Not sure if the test failures are related.

And just to confirm, have we removed all other places outside of KV where Ray writes to Redis?

**Comment by iycheng** [comment]
> LGTM on the API, based on the off-thread discussion. I did not check the moved code in `src/ray/gcs/store_client/redis_store_client.cc`. Not sure if the test failures are related.
> 
> 
> 
> And just to confirm, have we removed all other places outside of KV where Ray writes to Redis?

Let me check the test failure.

Yes, we moved given that we have run ray without Redis for a while already, where all data are written to memory. This only change the memory access. I also have another PR to use Redis backend without pip install Redis.

**Review by scv119 — APPROVED** [review]

**Pull request merged**